### PR TITLE
Cache packages to boost Container generator speed

### PR DESCRIPTION
**Pull request by belemaire**
This PR introduces a simple caching mechanism to cache packages used during Container generation (MiniApps, APIs & APIs implementations, native modules).

As a baseline, for a big sample production Container, if not changing anything in the Container but just calling `cauldron regen-container`, time to generate a Container goes from ~10min to ~6min (~40% decrease). This is the ideal scenario, with all packages versions being cached. 

Speed gains decrease linearly with the number of non-cached packages versions.

The cache has a default size of 2GB (for this big sample production Container, all packages at a specific version represent ~1GB, so this leaves us room to cache a few different versions). The package cache max size is configurable through `platform config max-package-cache-size`. Once max cache size is exceeded, the least recently accessed package(s) will be removed from the cache until total cache size gets below max cache size.

The package cache is disabled by default, and still experimental, so not yet documented. It will be shipped in Electrode Native 0.21.0, battle tested internally and will be moved out of experimental phase and documented in Electrode Native 0.22.0 most probably.

Enabling the package cache can be done through `platform config package-cache-enabled true`.

The packages cache directory is located inside the `.ern` folder in `packages-cache`.

Most of Container Generation time is now spent for composite MiniApp generation and bundling. This is a little bit trickier to cache but possible. That being said, we won't really be able to decrease bundling speed (if we stick to Metro) as it is out of our control. But caching composite prior to bundling could add a few minutes gain, probably making our reference container generation from ~10min down to ~3-4min.

### CODE REVIEW

File: ern-core/src/FsCache.ts [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+      await writeJSON(this.cacheManifestPath, manifestObj)
+      return cacheEntry.path
+    }
+  }
+
+  /**
+   * Gets the cache manifest object
+   */
+  private async getManifestObj(): Promise<CacheManifest> {
+    return readJSON(this.cacheManifestPath)
```
Comment (krunalsshah) [inline ern-core/src/FsCache.ts]: readJSONSync()?

File: ern-core/src/FsCache.ts [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  /**
+   * Time (epoc) when the entry was last accessed
+   */
+  lastAccessed: number
+}
+
+const cacheManifestFileName = 'cache-manifest.json'
+export const maxDefaultCacheSize = 2 * 1024 * 1024 * 1024 /* 2GB */
+
+type AddObjectToDirectoryFun<T> = (obj: T, dirPath: string) => Promise<void>
```
Comment (krunalsshah) [inline ern-core/src/FsCache.ts]: `AddObjectToDirectoryFun` is not clear , sorry :(

File: ern-core/src/FsCache.ts [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    await writeJSON(this.cacheManifestPath, manifestObj)
+    return pathToObjectCacheDir
+  }
+
+  /**
+   * Gets path to the directory containing a cached object
+   * @param obj The object for which to retrieve cached directory path
+   */
+  public async getObjectCachePath(obj: T): Promise<string | void> {
+    const manifestObj = await this.getManifestObj()
```
Comment (krunalsshah) [inline ern-core/src/FsCache.ts]: If it's possible to read `return readJSONSync(this.cacheManifestPath)`, we can `await writeJSONSync(this.cacheManifestPath, manifestObj)` (line 193) ? In which case we can avoid `await`

File: ern-core/src/FsCache.ts [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+      writeJSONSync(this.cacheManifestPath, { entries: [], size: 0 })
+    }
+  }
+
+  /**
+   * Indicates whether a given object is currently stored in the cache
+   * @param obj The object to check for presence in the cache
+   */
+  public async isInCache(obj: T): Promise<boolean> {
+    const manifestObj = await this.getManifestObj()
```
Comment (krunalsshah) [inline ern-core/src/FsCache.ts]: If it's possible for `getManifestObj` to `readJSONSync()` we an can eliminate need for `async`

File: ern-core/src/FsCache.ts [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+import { readJSON, writeJSON, writeJSONSync } from './fileUtil'
+import shell from './shell'
+import uuidv4 from 'uuid/v4'
+import getSize from 'get-folder-size'
+
+/**
+ * Gets the total size (in bytes) of a given directory
+ * @param dirPath path to the directory for which to compute size
+ */
+const getSizeAsync: (dirPath: string) => Promise<number> = dirPath =>
```
Comment (krunalsshah) [inline ern-core/src/FsCache.ts]: Treat this comment as informational, why did we have to read it into a `sync` block?

**Comment by krunalsshah** [comment]
👍 thanks for the optimization.

File: ern-core/src/FsCache.ts (hunk 1 above)
Comment (belemaire) [reply ern-core/src/FsCache.ts]: This is an `async` method, it returns a `Promise`. `readJson` returns a Promise so nothing wrong here.

File: ern-core/src/FsCache.ts (hunk 3 above)
Comment (belemaire) [reply ern-core/src/FsCache.ts]: We should use asynchronous methods as much as possible in the code base. 
Async (promise based) is the way to go everywhere doable.
Unfortunately, constructors cannot be executed asynchronously, thus there was no other choice but to choose sync version in the constructor (unless change of design with lazy init, but wanted to keep things simple). That's not a big issue.

File: ern-core/src/FsCache.ts (hunk 2 above)
Comment (belemaire) [reply ern-core/src/FsCache.ts]: Agreed, will rename to `AddObjectToCacheDirectory`

File: ern-core/src/FsCache.ts (hunk 4 above)
Comment (belemaire) [reply ern-core/src/FsCache.ts]: Same as previous comment.

**Comment by belemaire** [comment]
Pushed changes based on review :
- Rename `AddObjectToDirectoryFun` to `AddObjecToCacheDirectory`

**Pull request merged**